pkg/clients/zone/zone.go

```diff
@@ -0,0 +1,101 @@
+package zone
+
+import (
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/awserr"
+	"github.com/aws/aws-sdk-go-v2/service/route53"
+	"github.com/aws/aws-sdk-go-v2/service/route53/route53iface"
+	"github.com/google/uuid"
+
+	"github.com/crossplane/provider-aws/apis/network/v1alpha3"
+)
+
+// Client defines Route53 Client operations
+type Client interface {
+	CreateZoneRequest(cr *v1alpha3.Zone) route53.CreateHostedZoneRequest
+	DeleteZoneRequest(id *string) route53.DeleteHostedZoneRequest
+	GetZoneRequest(id string) route53.GetHostedZoneRequest
+	UpdateZoneRequest(id, comment *string) route53.UpdateHostedZoneCommentRequest
+}
+
+type zoneClient struct {
+	zone route53iface.ClientAPI
+}
+
+// NewClient creates new AWS Client with provided AWS Configurations/Credentials
+func NewClient(config *aws.Config) Client {
+	return &zoneClient{zone: route53.New(*config)}
+}
+
+// GetZoneRequest returns a route53 GetHostedZoneRequest using which a route53
+// Hosted Zone can be fetched and checked for existence.
+func (c *zoneClient) GetZoneRequest(id string) route53.GetHostedZoneRequest {
+	return c.zone.GetHostedZoneRequest(&route53.GetHostedZoneInput{
+		Id: &id,
+	})
+}
+
+// CreateZoneRequest returns a route53 CreateHostedZoneRequest using which a route53
+// Hosted Zone can be created.
+func (c *zoneClient) CreateZoneRequest(cr *v1alpha3.Zone) route53.CreateHostedZoneRequest {
+	if cr.Spec.ForProvider.CallerReference == nil {
+		cr.Spec.ForProvider.CallerReference = getCallerRef()
```

Comment:
```
	// A unique string that identifies the request and that allows failed CreateHostedZone
	// requests to be retried without the risk of executing the operation twice.
	// You must use a unique CallerReference string every time you submit a CreateHostedZone
	// request. CallerReference can be any unique string, for example, a date/time
	// stamp.
	//
	// CallerReference is a required field
```

I believe we can use `metadata.resourceVersion` for this field so that we don't call `Create` twice with exactly same CR.

Reply:
`metadata.resourceVersion` won't work here as it keeps on changes with every update to the object hence I have used `metadata.UID` instead.